design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,483 @@
+but it may be omitted when authoring an ApplicationConfiguration. Metadata such
+as names and labels may be filled automatically by the OAM runtime if a distinct
+instance of a trait configuration is created as part of processing an
+ApplicationConfiguration. As with workloads, Kubernetes runtimes may treat
+traits as an [embedded object]. Each trait resource is embedded beneath the key
+`trait`. This allows the spec to add other keys alongside `trait` in future
+without breaking compatibility.
+
+### Scopes
+
+This document proposes that application scopes, while maintaining their current
+purpose of grouping application components, function identically to traits from
+an API perspective. Available kinds of scope are defined via a ScopeDefinition,
+which also specifies whether a component may exist in multiple scopes of a
+particular kind:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ScopeDefinition
+metadata:
+  name: networkscope.core.oam.dev
+spec:
+  allowComponentOverlap: true
+  definitionRef:
+    name: networkscope.core.oam.dev
+```
+
+The above ScopeDefinition references a Kubernetes CustomResourceDefinition named
+networkscope.core.oam.dev, designating custom resources defined by said CRD to
+be valid OAM scopes. As with the other OAM definition resources OAM runtimes
+that are not powered by Kubernetes could opt to either implement the CRD
+concept, or introduce a new field to the ScopeDefinition spec that defines a
+kind of scope and its schema. Application operators can discover available OAM
+scope kinds by running `kubectl get scopedefinitions`. Scopes would most
+frequently be defined by OAM runtimes or infrastructure operators.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: NetworkScope
+metadata:
+  name: example-vpc-network
+  labels:
+    region: us-west
+    environment: production
+spec:
+  networkId: cool-vpc-network
+  subnetIds:
+  - cool-subnetwork
+  - cooler-subnetwork
+  - coolest-subnetwork
+  internetGatewayType: nat
+```
+
+Scopes would be configured by creating an instance of configuration - a scope
+resource that may be referenced by ApplicationConfigurations. This is not
+functionally dissimilar to the contemporary OAM specification, which uses an
+ApplicationConfiguration to create an instance of scope configuration that is
+referenced by other ApplicationConfigurations.
+
+### Application Configuration
+
+Finally, the ApplicationConfiguration resource ties together a set of
+components, traits, and scopes as it does today:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ApplicationConfiguration
+metadata:
+  name: cool-example
+spec:
+  components:
+  - componentName: example-server
+    parameterValues:
+    - name: instanceName
+      value: cool-example
+    - name: cacheSecret
+      value: cache-connection
+    traits:
+    - trait:
+        apiVersion: core.oam.dev/v1alpha1
+        kind: ManualScalerTrait
+        spec:
+          replicaCount: 3
+    scopes:
+    - scopeRef:
```

Comment:
or just `scope`?

Reply:
yeah, align with trait

Reply:
I think this should stay as `scopeRef`. It's a _reference to a scope_, whereas a trait is _an embedded trait_.  Per API conventions (I feel like a broken record referring to these conventions) fields that are a reference to a `kind: Foo` should be named `fooRef`.

https://github.com/kubernetes/community/blob/96203/contributors/devel/sig-architecture/api-conventions.md#naming-conventions